### ISSUE TITLE
Broken libreSSL tls_conn_change_cert implementation

### ISSUE THREAD

**Issue by sreimers** [baresip/re#1135]
Looks like there are some broken libreSSL implementations https://github.com/baresip/re/pull/1127.

Let's start with `test_tls_cli_conn_change_cert` and `tls_conn_change_cert`, this does not work within libressl since `SSL_certs_clear` is not supported.

```c
tlstest: TEST_STRCMP: /home/sreimers/projekte/baresip/baresip-linux/re/test/tls.c:539: failed
expected string: (21 bytes)
"Mr Retest Client Cert"
actual string: (9 bytes)
"127.0.0.1"
```

I wonder why the ssl object is changed and not the ssl ctx before? Since within `tls_start_tcp` a new ssl object is created from ctx. This way `SSL_certs_clear` can be avoided. 

https://github.com/baresip/re/blob/b41f503085af62e1d7f1d12b1226ee361b7e02cc/src/sip/transp.c#L827-L843

We should keep all SSL implementations very generic.

@cHuberCoffee

**Comment by sreimers** [comment]
Maybe `SSL_set_SSL_CTX` can be used here too, will try after https://github.com/baresip/re/pull/1136 is merged.

**Comment by cHuberCoffee** [comment]
So the allocation of the `tls struct` containing an `SSL_CTX` is done by `uag_transp_add`.
Basically we only have one `SSL_CTX` where all `SSL objects` are inherited. This includes the loaded certificate which we need to replace. The `SSL_certs_clear` is for sure not a good implementation for doing so. The `SSL_set_SSL_CTX` would mean a new context for each certificate configured on user agent level.

It would be possible to create a new context for each such a configuration and saving these contexts as hash_list instead of the certificate path. Then using the `SSL_set_SSL_CTX`. For OpenSSL we have to find the same functionality. At least in the documentation i found nothing.

**Comment by sreimers** [comment]
`SSL_set_SSL_CTX` is supported and recommended by OpenSSL but not very documented:

https://github.com/baresip/re/pull/1136

**Comment by cHuberCoffee** [comment]
"not very documented" points it out.
But you are right, following down the implementation in OpenSSL for the function does look a lot like its exactly what we want to archive here.

**Comment by cspiel1** [comment]
Is v3.14.0 okay for this issue? Today we planned to release v3.13.0.

**Comment by maximilianfridrich** [comment]
@sreimers Would something like the following be an option? In `src/tls/openssl/tls_tcp.c:331`:

```
#if !defined(LIBRESSL_VERSION_NUMBER)
	SSL_certs_clear(tc->ssl);
#else
	(void)SSL_use_certificate(tc->ssl, NULL);
	(void)SSL_use_PrivateKey(tc->ssl, NULL);
	SSL_clear_extra_chain_certs(SSL_get_SSL_CTX(tc->ssl));
#endif
```

**Comment by maximilianfridrich** [comment]
Nevermind, LibreSSL also does a NULL check on the second parameter.

**Comment by alfredh** [comment]
Is there any interest in keeping support for LibreSSL ?
At the moment there is at least 3-4 functions that are broken with LibreSSL.

LibreSSL is default on OpenBSD, and can be installed on platforms
like FreeBSD, macOS, etc ..

**Comment by cspiel1** [comment]
Regarding to the changes of @cHuberCoffee I have to mention that we don't use LibreSSL. Currently we fully depend on OpenSSL. From our point of view it would simplify the code and unit tests to remove the support for LibreSSL. Thus, it's a good question: **Is LibreSSL support still needed?**

**Comment by sreimers** [comment]
I don't think we should drop support for LibreSSL. This way we can keep the TLS API more generic. Not all features supported by OpenSSL are widely implemented for all clients (like [TLS 1.3 Post Handshake Auth](https://github.com/baresip/re/pull/1323)) or used.
This makes it much easier to adapt other TLS stacks in the future, such as WolfSSL or Mbed TLS, AWS-LC (https://www.haproxy.com/blog/state-of-ssl-stacks).

Most of the time there is a story behind an unsupported LibreSSL Api (security, best practice, rarely used, etc.).

> From our point of view, it would simplify the code and unit tests to remove support for LibreSSL.

There is no more LibreSSL specific code that could be cleaned up or simplified (other than this issue).

**Comment by alfredh** [comment]
The advantage of supporting LibreSSL is that we get less dependency on OpenSSL.
If we dont support LibreSSL we will be 100% dependent on OpenSSL.

The disadvantage is slightly increased maintenance overhead ...
